# Taller Integrador — Módulo 1
## Máster en IA & Data Science | DevSeniorCode

**Nombre del estudiante:** Sergio Andres Zapata Ruiz  
**Fecha de entrega:** ___________________________  
**Repositorio GitHub:** ___________________________

---

> Este taller evalúa todo lo visto en el Módulo 1:  
> Python · Pandas · Git & GitHub · Documentación

### ¿Cómo entregar?
1. Completa todas las celdas marcadas con 
2. Asegúrate de que el notebook corra sin errores (Kernel → Restart & Run All)
3. Sube el notebook y el CSV a tu repositorio de GitHub
4. Incluye un README.md con tu nombre y una descripción breve del proyecto

---

## Contexto del Dataset

Trabajarás con el dataset **`salud_mundial.csv`**, que contiene indicadores de salud de **159 países**.

| Columna | Descripción |
|---|---|
| `pais` | Nombre del país |
| `region` | Continente o región geográfica |
| `nivel_ingresos` | Nivel económico: Low, Medium, High, Very High |
| `esperanza_vida` | Esperanza de vida al nacer (años) |
| `gasto_salud_usd` | Gasto per cápita en salud (USD/año) |
| `mortalidad_infantil` | Muertes por cada 1,000 nacidos vivos |
| `medicos_por_1000` | Médicos por cada 1,000 habitantes |
| `poblacion_urbana_pct` | % de población que vive en zonas urbanas |
| `obesidad_pct` | % de la población con obesidad |
| `diabetes_pct` | % de la población con diabetes |
| `tabaquismo_pct` | % de fumadores |
| `camas_hospital_por_1000` | Camas hospitalarias por 1,000 hab. |
| `vacunacion_pct` | % de cobertura de vacunación |

---

---
## Parte 1 — Python Fundamental
### (30 puntos)

Antes de usar Pandas, demostraremos que dominas las estructuras básicas de Python.

---

### 1.1 — Listas y operaciones básicas

Tienes la siguiente lista con la esperanza de vida de 10 países (en años).

In [1]:
# Lista dada — NO modificar
esperanzas_de_vida = [72.3, 81.5, 60.1, 78.9, 55.4, 83.2, 67.8, 74.0, 49.3, 79.6]

**Ejercicio 1.1** — Usando Python puro (sin Pandas ni funciones built-in como `sum()` o `max()`):
- Calcula el **promedio** de la lista usando un bucle `for`
- Encuentra el **valor máximo** usando un bucle `for`
- Encuentra el **valor mínimo** usando un bucle `for`

In [14]:
# Tu código aquí
# Promedio
suma = 0
cont = 0
maximo = esperanzas_de_vida[0]
minimo = esperanzas_de_vida[0]

for numero in esperanzas_de_vida:
    suma += numero
    cont +=1

    if numero > maximo: # analizar numero maximo
        maximo = numero

    if numero < minimo: # analizar numero minimo
        minimo = numero
promedio = suma / cont
print(f"El promedio de la lista 'esperanzas_de_vida' es {promedio}")
  
# Máximo
print(f"El numero mayor de la lista 'esperanzas_de_vida' es {maximo}")

# Mínimo
print(f"El mumero menor de la lista 'esperanzas_de_vida' es {minimo}")



El promedio de la lista 'esperanzas_de_vida' es 70.21000000000001
El numero mayor de la lista 'esperanzas_de_vida' es 83.2
El mumero menor de la lista 'esperanzas_de_vida' es 49.3


## Respuesta:
- `Promedio =`  70.21000000000001
- `Maximo =` 83.2
- `Minimo =` 49.3

### 1.2 — Diccionarios

Los diccionarios son fundamentales en Data Science para representar registros.

**Ejercicio 1.2** — Crea un diccionario llamado `pais_info` que represente a Colombia con los siguientes campos:
- `nombre`: "Colombia"
- `region`: "Americas"  
- `esperanza_vida`: 76.5
- `gasto_salud_usd`: 560.3
- `nivel_ingresos`: "High"

Luego:
1. Imprime solo las **claves** del diccionario
2. Imprime solo los **valores**
3. Agrega un nuevo campo: `vacunacion_pct` con valor 82.1
4. Imprime el diccionario completo

In [3]:
# Tu código aquí
pais_info = {
    "nombre": "Colombia",
    "region": "Americas", 
    "esperanza_vida": 76.5,
    "gasto_salud_usd": 560.3,
    "nivel_ingresos": "High"
}

print("Claves: ")
for clave in pais_info.keys():
    print("-", clave)

print("Valores: ")
for valor in pais_info.values():
    print("-", valor)

pais_info.setdefault("vacunacion_pct",82.1) # solo se agrega si la clave no existe "vacunacion_pct"

print(f"Diccionario completo:\n- {pais_info}")

Claves: 
- nombre
- region
- esperanza_vida
- gasto_salud_usd
- nivel_ingresos
Valores: 
- Colombia
- Americas
- 76.5
- 560.3
- High
Diccionario completo:
- {'nombre': 'Colombia', 'region': 'Americas', 'esperanza_vida': 76.5, 'gasto_salud_usd': 560.3, 'nivel_ingresos': 'High', 'vacunacion_pct': 82.1}


## Respuesta:
### Claves:
- `nombre`
- `region`
- `esperanza_vida`
- `gasto_salud_usd`
- `nivel_ingresos`
### Valores:
- `Colombia`
- `Americas`
- `76.5`
- `560.3`
- `High`
### Diccionario completo:
- {'nombre': 'Colombia', 'region': 'Americas', 'esperanza_vida': 76.5, 'gasto_salud_usd': 560.3, 'nivel_ingresos': 'High', 'vacunacion_pct': 82.1}

### 1.3 — Funciones

Las funciones nos permiten reutilizar lógica.

**Ejercicio 1.3** — Crea una función llamada `clasificar_esperanza_vida` que reciba un número (años) y retorne:
- `"Alta"` si es >= 75
- `"Media"` si es entre 60 y 74.9
- `"Baja"` si es < 60

Prueba la función con al menos 5 valores distintos e imprime los resultados.

In [4]:
# Tu código aquí
def clasificar_esperanza_vida(years:float)->str:
    if years >= 75:
        return "Alta"
    elif years >= 60 and years <= 74.9:
        return "Media"
    elif years < 60:
        return "Baja"
    return ""

print(clasificar_esperanza_vida(75))
print(clasificar_esperanza_vida(75.78))
print(clasificar_esperanza_vida(65))
print(clasificar_esperanza_vida(30.4))
print(clasificar_esperanza_vida(-19.5))



Alta
Alta
Media
Baja
Baja


### 1.4 — Listas + Funciones + Bucles (integrador)

**Ejercicio 1.4** — Usa la función creada en 1.3 y la lista del ejercicio 1.1 para:
1. Crear una **nueva lista** llamada `clasificaciones` que contenga la clasificación de cada esperanza de vida
2. Imprimir cada par: `(valor, clasificación)` usando un bucle `for`

In [5]:
# Tu código aquí
clasificaciones = [(numero, clasificar_esperanza_vida(numero)) for numero in esperanzas_de_vida]

for numero, esperansa_vida in clasificaciones:
    print(f"para un numero '{numero}' se considera una esperanza de vida '{esperansa_vida}'")



para un numero '72.3' se considera una esperanza de vida 'Media'
para un numero '81.5' se considera una esperanza de vida 'Alta'
para un numero '60.1' se considera una esperanza de vida 'Media'
para un numero '78.9' se considera una esperanza de vida 'Alta'
para un numero '55.4' se considera una esperanza de vida 'Baja'
para un numero '83.2' se considera una esperanza de vida 'Alta'
para un numero '67.8' se considera una esperanza de vida 'Media'
para un numero '74.0' se considera una esperanza de vida 'Media'
para un numero '49.3' se considera una esperanza de vida 'Baja'
para un numero '79.6' se considera una esperanza de vida 'Alta'


---
## Parte 2 — Análisis con Pandas
### (50 puntos)

---

### 2.1 — Carga y exploración inicial

🔧 **Ejercicio 2.1** — Importa Pandas y carga el archivo `salud_mundial.csv`.  
Luego ejecuta los comandos necesarios para responder:
1. ¿Cuántas filas y columnas tiene el dataset?
2. ¿Qué tipos de datos tiene cada columna?
3. ¿Cuántos valores nulos hay por columna?
4. Muestra las primeras 8 filas

In [15]:
# Tu código aquí
import pandas as pd

# Cargar el dataset
df_salud_mundial = pd.read_csv("salud_mundial.csv")
# df_salud_mundial.head()

# Dimensiones
filas, columnas = df_salud_mundial.shape
print("Dimenciones:")
print(f"El df tiene {filas} filas")
print(f"El df tiene {columnas} columnas")


# Tipos de datos
columnas_tipo = [(columna, type(df_salud_mundial[columna][0])) for columna in df_salud_mundial.columns]
print("\nTipo de dato por columna:")
for columa, tipo in columnas_tipo:
    print(f"- {columa} : {tipo}")

# Valores nulos
print("\nValores nulos por columna: ('columna', 'suma de nulos')")
conteo_nulos = [(pais, suma_nulos) for pais, suma_nulos in df_salud_mundial.isnull().sum().items()]
# separamos para dar una mejor visual
print(conteo_nulos[:5])
print(conteo_nulos[5:9])
print(conteo_nulos[9:])


# Primeras 8 filas
# en la siguiente celda de codigo para mejor visual


Dimenciones:
El df tiene 159 filas
El df tiene 13 columnas

Tipo de dato por columna:
- pais : <class 'str'>
- region : <class 'str'>
- nivel_ingresos : <class 'str'>
- esperanza_vida : <class 'numpy.float64'>
- gasto_salud_usd : <class 'numpy.float64'>
- mortalidad_infantil : <class 'numpy.float64'>
- medicos_por_1000 : <class 'numpy.float64'>
- poblacion_urbana_pct : <class 'numpy.float64'>
- obesidad_pct : <class 'numpy.float64'>
- diabetes_pct : <class 'numpy.float64'>
- tabaquismo_pct : <class 'numpy.float64'>
- camas_hospital_por_1000 : <class 'numpy.float64'>
- vacunacion_pct : <class 'numpy.float64'>

Valores nulos por columna: ('columna', 'suma de nulos')
[('pais', 0), ('region', 0), ('nivel_ingresos', 0), ('esperanza_vida', 4), ('gasto_salud_usd', 5)]
[('mortalidad_infantil', 10), ('medicos_por_1000', 9), ('poblacion_urbana_pct', 0), ('obesidad_pct', 3)]
[('diabetes_pct', 7), ('tabaquismo_pct', 4), ('camas_hospital_por_1000', 7), ('vacunacion_pct', 4)]


In [7]:
# Primeras 8 filas
df_salud_mundial.head(8)

,pais,region,nivel_ingresos,esperanza_vida,gasto_salud_usd,mortalidad_infantil,medicos_por_1000,poblacion_urbana_pct,obesidad_pct,diabetes_pct,tabaquismo_pct,camas_hospital_por_1000,vacunacion_pct
0,Afghanistan,Asia,Low,60.9,18.4,NaN,0.19,40.8,15.2,9.2,10.1,1.0,41.1
1,Albania,Europe,Medium,65.1,673.8,28.1,1.18,45.0,27.2,9.7,17.3,1.2,84.5
2,Algeria,Africa,Medium,74.5,525.3,28.2,0.59,47.3,15.2,7.6,20.8,1.3,71.4
3,Angola,Africa,Low,NaN,37.1,54.7,0.79,37.4,13.4,7.8,28.2,1.6,48.7
4,Argentina,Americas,High,79.4,1809.5,7.6,NaN,71.5,18.9,10.1,33.0,3.4,81.7
5,Armenia,Europe,Medium,69.2,797.5,24.2,2.44,67.5,10.2,12.8,32.0,2.3,71.1
6,Australia,Oceania,Very High,NaN,9964.1,NaN,4.42,87.2,20.6,10.3,19.7,7.3,93.8
7,Austria,Europe,Very High,83.7,3685.2,3.9,2.71,91.0,31.0,5.9,18.6,6.0,90.9


### 2.2 — Limpieza de datos

🔧 **Ejercicio 2.2** — El dataset tiene algunos valores nulos (¡datos reales siempre tienen este problema!).  
1. Rellena los valores nulos de columnas **numéricas** con la **media** de cada columna
2. Verifica que ya no queden nulos
3. Explica en una celda de texto: ¿Por qué usamos la media y no simplemente eliminamos esas filas?

In [16]:
# Tu código aquí
def obtener_media(serie:pd.Series)->float:
    suma = 0
    cantidad_datos = len(serie)
    for dato in serie:
        if pd.notna(dato):
            suma += dato
    return suma / cantidad_datos


lista_columnas_con_nulos = conteo_nulos[3:] # eliminamos columnas str
# separamos columnas con nulos
lista_columnas_con_nulos = [columna for columna, num_nulos in lista_columnas_con_nulos if num_nulos > 0]

for columa in lista_columnas_con_nulos:
    # media_columna = obtener_media(df_salud_mundial[columa]) # usando la funcion que construimos
    media_columna = df_salud_mundial[columa].median()
    df_salud_mundial[columa] = df_salud_mundial[columa].fillna(media_columna)

# df_salud_mundial.head(8)
print(df_salud_mundial.isnull().sum())



pais                       0
region                     0
nivel_ingresos             0
esperanza_vida             0
gasto_salud_usd            0
mortalidad_infantil        0
medicos_por_1000           0
poblacion_urbana_pct       0
obesidad_pct               0
diabetes_pct               0
tabaquismo_pct             0
camas_hospital_por_1000    0
vacunacion_pct             0
dtype: int64


**Responde aquí** (doble clic para editar):

*¿Por qué usamos la media para rellenar nulos en vez de eliminar las filas?*

> Tu respuesta:

### 2.3 — Estadísticas descriptivas

🔧 **Ejercicio 2.3**
1. Muestra el resumen estadístico completo del dataset
2. ¿Cuál es la esperanza de vida **promedio** mundial?
3. ¿Cuál es el país con **mayor** gasto en salud? ¿Y el de **menor**?
4. ¿Cuál es la **mediana** de la mortalidad infantil?

In [17]:
# Tu código aquí
print(f"La esperanza de vida promedio mundial es: {df_salud_mundial['esperanza_vida'].median()} años")

df_salud_mundial_gasto_dalud = df_salud_mundial.groupby('pais')['gasto_salud_usd']
lista_gasto = list(df_salud_mundial_gasto_dalud.mean().items())

pais_mas_gaston = lista_gasto[0][0]
mayor_gasto = lista_gasto[0][1]
pais_menos_gaston = lista_gasto[0][0]
menor_gasto = lista_gasto[0][1]

for pais, gasto in lista_gasto:
    if gasto > mayor_gasto:
        pais_mas_gaston = pais
        mayor_gasto = gasto
    if gasto < menor_gasto:
        pais_menos_gaston = pais
        menor_gasto = gasto
print(f"El pais con mayor gasto en salud es: {pais_mas_gaston} y gasta: {mayor_gasto}")
print(f"El pais con menor gasto en salud es: {pais_menos_gaston} y gasta: {menor_gasto}")




La esperanza de vida promedio mundial es: 73.2 años
El pais con mayor gasto en salud es: Slovakia y gasta: 10311.8
El pais con menor gasto en salud es: Afghanistan y gasta: 18.4


### 2.4 — Filtrado y selección

🔧 **Ejercicio 2.4**
1. Filtra los países de **América** (`region == "Americas"`) y muestra solo las columnas: `pais`, `esperanza_vida`, `gasto_salud_usd`
2. Filtra los países con esperanza de vida **mayor a 80 años** — ¿cuántos son? ¿De qué regiones son?
3. Filtra los países con **nivel_ingresos == "Low"** y calcula su promedio de `mortalidad_infantil`
4. ¿Cuántos países tienen vacunación por encima del 90%?

In [30]:
# Tu código aquí
df_salud_mundial
df_paises_americas=df_salud_mundial.loc[df_salud_mundial["region"]=="Americas"]
lista_americas=["pais","esperanza_vida","gasto_salud_usd"]
df_paises_americas_filtrados=pd.DataFrame()
for columna in lista_americas:
    df_paises_americas_filtrados[columna]=df_paises_americas[columna]
df_paises_americas_filtrados


,pais,esperanza_vida,gasto_salud_usd
4,Argentina,79.4,1809.50
11,Bolivia,75.8,322.40
12,Brazil,74.1,2913.70
16,Canada,78.9,4019.60
17,Chile,79.2,854.50
19,Colombia,75.2,2133.70
21,Costa Rica,76.3,1772.80
23,Cuba,78.9,1804.20
26,Dominican Republic,73.2,591.40
27,Ecuador,69.2,638.80


2. Filtra los países con esperanza de vida **mayor a 80 años** — ¿cuántos son? ¿De qué regiones son?

In [47]:
df_salud_mundial
df_esperanza_mayor80=df_salud_mundial.loc[df_salud_mundial["esperanza_vida"]>80]
filas,columnas=df_esperanza_mayor80.shape
print(f"La cantidad de paises que tienen una esperanza de vida mayor a 80 son: {filas}")
las_regiones=[]
for index,region in df_esperanza_mayor80["region"].items():
    if region not in las_regiones: 
        las_regiones.append(region)
print(f"Las regiones son: {las_regiones}")

La cantidad de paises que tienen una esperanza de vida mayor a 80 son: 23
Las regiones son: ['Europe', 'Asia', 'Americas']


3. Filtra los países con **nivel_ingresos == "Low"** y calcula su promedio de `mortalidad_infantil`

In [52]:
df_paises_low=df_salud_mundial.loc[df_salud_mundial["nivel_ingresos"]=="Low"]
promedio_mortalidad_infantil = df_paises_low["mortalidad_infantil"].mean()
print(f"El Promedio de la mortalidad infantil en los paises low es: {promedio_mortalidad_infantil}")


El Promedio de la mortalidad infantil en los paises low es: 62.556521739130446


4. ¿Cuántos países tienen vacunación por encima del 90%?

In [63]:
df_prom_vacunacion=df_salud_mundial.loc[df_salud_mundial["vacunacion_pct"]>90]
filas, columnas = df_prom_vacunacion.shape
print(f"La cantidad de paises vacunacion por encima del 90% son: {filas}")


La cantidad de paises vacunacion por encima del 90% son: 39


### 2.5 — Agrupaciones

🔧 **Ejercicio 2.5** — Usa `groupby()` para:
1. Calcular la **esperanza de vida promedio** por `region`
2. Calcular el **gasto en salud promedio** por `nivel_ingresos`
3. Calcular el **número de países** por región
4. ¿Qué región tiene la mayor esperanza de vida promedio? ¿Y la menor?

In [11]:
# Tu código aquí



### 2.6 — Nueva columna calculada

🔧 **Ejercicio 2.6**
1. Crea una nueva columna llamada `categoria_vida` aplicando la función `clasificar_esperanza_vida` que creaste en la Parte 1 a la columna `esperanza_vida`
2. ¿Cuántos países caen en cada categoría? (usa `value_counts()`)
3. ¿Qué porcentaje del total representa cada categoría?

In [12]:
# Tu código aquí



### 2.7 — Análisis libre (desafío)

**Ejercicio 2.7 (Opcional — puntos extra)** — Formula **una pregunta propia** sobre el dataset y respóndela con código.  

Ejemplos de preguntas que podrías hacer:
- ¿Existe relación entre el gasto en salud y la esperanza de vida?
- ¿Los países con más médicos tienen menor mortalidad infantil?
- ¿Qué región tiene mayor tasa de obesidad?

**Escribe tu pregunta:**

In [13]:
# Tu código aquí



---
## Parte 3 — Git & GitHub
### (20 puntos)

Esta parte se evalúa por el **estado real de tu repositorio** en GitHub.  
Documenta los comandos que usaste respondiendo las preguntas en las celdas de texto.

---

### 3.1 — Configuración del repositorio

**Responde** (doble clic para editar):

**¿Qué comandos usaste para crear e inicializar tu repositorio local?**
```bash
# Tu respuesta:

```

**¿Cómo conectaste tu repositorio local con GitHub?**
```bash
# Tu respuesta:

```

### 3.2 — Flujo de trabajo

**Responde:**

**¿Qué comando usaste para añadir los archivos al staging area?**
```bash
# Tu respuesta:

```

**¿Qué mensaje de commit escribiste? ¿Por qué elegiste ese mensaje?**
```bash
# Tu respuesta:

```

**¿Cómo subiste los cambios a GitHub?**
```bash
# Tu respuesta:

```

### 3.3 — Buenas prácticas

**Responde:**

**¿Qué archivos debería incluir un `.gitignore` en un proyecto de Data Science? Menciona al menos 3 y explica por qué.**

> Tu respuesta:

**¿Cuál es la diferencia entre `git add .` y `git add nombre_archivo`?**

> Tu respuesta:

---
## Parte 4 — Documentación (README)
### (Evaluada en GitHub)

Tu repositorio debe tener un archivo `README.md` que incluya:

- [ ] Título del proyecto
- [ ] Tu nombre
- [ ] Descripción del dataset (qué contiene, cuántos países, fuente)
- [ ] Cómo ejecutar el notebook (instrucciones paso a paso)
- [ ] Al menos 3 hallazgos o conclusiones del análisis
- [ ] Tecnologías usadas (Python, Pandas, Jupyter, Git)

---

### Checklist de entrega

Antes de hacer tu commit final, verifica:

- [ ] El notebook corre completo sin errores (Kernel → Restart & Run All)
- [ ] Todas las celdas están completadas
- [ ] Las celdas de texto tienen tus respuestas
- [ ] El archivo `salud_mundial.csv` está en el repositorio
- [ ] El `README.md` está completo
- [ ] El link al repositorio está en la celda de portada

---

## Criterios de evaluación

| Sección | Puntos |
|---|---|
| Parte 1 — Python | 30 |
| Parte 2 — Pandas | 50 |
| Parte 3 — Git | 20 |
| README en GitHub | 10 |
| Ejercicio libre 2.7 | +10 |
| **Total** | **110** |

